src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,489 @@
+        {
+            const fault_index =
+                (queue.faulty_blocks_repair_index + fault_offset) % faults_total;
+
+            const fault = &fault_data[fault_index];
+            if (fault.state == .waiting) {
+                requests[requests_count] = .{
+                    .block_address = fault_addresses[fault_index],
+                    .block_checksum = fault_data[fault_index].checksum,
+                };
+                requests_count += 1;
+            }
+        }
+
+        queue.faulty_blocks_repair_index =
+            (queue.faulty_blocks_repair_index + fault_offset) % faults_total;
+        return requests_count;
+    }
+
+    /// Count the number of *non-table* block repairs available.
+    pub fn request_blocks_available(queue: *const GridRepairQueue) usize {
+        assert(!queue.canceling);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        const faulty_blocks_free =
+            queue.faulty_blocks.capacity() -
+            queue.faulty_blocks.count() -
+            queue.options.tables_max * lsm_table_content_blocks_max;
+        return faulty_blocks_free;
+    }
+
+    /// Queue a faulty block to request from the cluster and repair.
+    pub fn request_block(queue: *GridRepairQueue, address: u64, checksum: u128) void {
+        assert(!queue.canceling);
+        assert(queue.request_blocks_available() > 0);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        if (queue.faulty_blocks.getPtr(address)) |fault| {
+            assert(fault.checksum == checksum);
+            assert(fault.state != .aborting);
+            return;
+        }
+
+        queue.faulty_blocks.putAssumeCapacityNoClobber(address, .{
+            .checksum = checksum,
+            .progress = .block,
+        });
+    }
+
+    pub fn request_table_queued(queue: *GridRepairQueue, address: u64, checksum: u128) bool {
+        assert(!queue.canceling);
+        maybe(queue.request_blocks_available() == 0);
+        assert(queue.faulty_tables.count <= queue.options.tables_max);
+
+        var tables = queue.faulty_tables.peek();
+        while (tables) |queue_table| : (tables = queue_table.next) {
+            if (queue_table.index_address == address) {
+                assert(queue_table.index_checksum == checksum);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    pub fn request_table(
+        queue: *GridRepairQueue,
+        callback: *const fn (*RepairTable, RepairTableResult) void,
+        table: *RepairTable,
+        address: u64,
+        checksum: u128,
+    ) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_tables.count < queue.options.tables_max);
+        assert(!queue.request_table_queued(address, checksum));
+
+        var tables = queue.faulty_tables.peek();
+        while (tables) |queue_table| : (tables = queue_table.next) assert(queue_table != table);
```

Comment:
Hm, this `while` just inlines the assert above?

Reply:
Not quite:

- `queue.request_table_queued()` checks whether we have already queued that particular `address` and `checksum`.
- The `while` loop checks that we don't accidentally (illegally) reuse the `table: *RepairTable` context object. (E.g. if, somehow, the `Replica` released the table from its `IOPS` prematurely and tried to reuse it.)